batch.go

```diff
@@ -0,0 +1,94 @@
+package iavl
+
+import (
+	dbm "github.com/cosmos/cosmos-db"
+)
+
+// BatchWithFlusher is a wraper
+// around batch that flushes batch's data to disk
+// as soon as the configurable limit is reached.
+type BatchWithFlusher struct {
+	db                      dbm.DB    // This is only used to create new batch
+	batch                   dbm.Batch // Batched writing buffer.
+	batchSizeFlushThreshold int       // The maximum size of the batch in bytes before it gets flushed to disk
+}
+
+var _ dbm.Batch = BatchWithFlusher{}
+
+// this value is used to estimate the additional size when adding an entry to a batch
+// additional size = key length + value length + possibleEntryOverHead
+var possibleEntryOverHead = 100
+
+// Ethereum has found that commit of 100KB is optimal
+var defaultFlushThreshold = 100000
+
+// NewBatchWithFlusher returns new BatchWithFlusher wrapping the passed in batch
+func NewBatchWithFlusher(db dbm.DB, flushThreshold int) BatchWithFlusher {
+	return BatchWithFlusher{
+		db:                      db,
+		batch:                   db.NewBatchWithSize(flushThreshold),
+		batchSizeFlushThreshold: flushThreshold,
+	}
+}
+
+// Set sets value at the given key to the db.
+// If the set causes the underlying batch size to exceed batchSizeFlushThreshold would be reached,
+// the batch is flushed to disk, cleared, and a new one is created with buffer pre-allocated to threshold.
+// The addition entry is then added to the batch.
+func (b BatchWithFlusher) Set(key []byte, value []byte) error {
+	size, err := b.batch.GetByteSize()
+	if err != nil {
+		return err
+	}
+	if size+len(key)+len(value)+possibleEntryOverHead >= b.batchSizeFlushThreshold {
+		err = b.batch.Write()
+		if err != nil {
+			return err
+		}
+		b.batch = b.db.NewBatchWithSize(b.batchSizeFlushThreshold)
+	}
+	err = b.batch.Set(key, value)
```

Comment:
What do you think about changing this to the following flow:
1. b.batch.Set(key, value)
2. b.batch.GetByteSize()
3. if batch size >= b.batchSizeFlushThreshold, flush it to disk

This way, we can be agnostic of the `possibleEntryOverHead` that is db backend specific.

I think that exceeding `b.batchSizeFlushThreshold` is acceptable. We can just redefine this method to flush whenever the threshold is actually exceeded

Reply:
I just don't want to re-allocate the content of the batch when we exceed `b.batchSizeFlushThreshold`, since we've already pre-allocated the batch with `NewBatchWithSize`

Reply:
Why do you think that this would require redundant re-allocation? 

It seems to me that we only need to re-allocate after we flush the batch exceeding the threshold which is the behavior in the current design as well

Reply:
it's because we specify a limit for the batch, we pre-allocate the batch with that limit size, and so when we unknowingly exceed that limit, (for example we set a key/value that make the batch size = limit + 10 byte ), the batch has to be allocated again since it is out of capacity

Reply:
Makes sense. Thank you